// @workspace/.ignored_api-client-react/src/generated/api.ts
/**
 * Generated by orval v8.5.3 🍺
 * Do not edit manually.
 * Api
 * Early Dementia Detection via Speech Analysis API
 * OpenAPI spec version: 0.1.0
 */
import { useMutation, useQuery } from "@tanstack/react-query";
import type {
  MutationFunction,
  QueryFunction,
  QueryKey,
  UseMutationOptions,
  UseMutationResult,
  UseQueryOptions,
  UseQueryResult,
} from "@tanstack/react-query";

import type {
  AnalysisResult,
  AnalyzeTextBody,
  ChatbotMessageBody,
  ChatbotResponse,
  CreateReportBody,
  GetReportsParams,
  HealthStatus,
  Report,
  TranscribeAudioBody,
} from "./api.schemas";

import { customFetch } from "../custom-fetch";
import type { ErrorType, BodyType } from "../custom-fetch";

type AwaitedInput<T> = PromiseLike<T> | T;

type Awaited<O> = O extends AwaitedInput<infer T> ? T : never;

type SecondParameter<T extends (...args: never) => unknown> = Parameters<T>[1];

/**
 * Returns server health status
 * @summary Health check
 */
export const getHealthCheckUrl = () => {
  return `/api/healthz`;
};

export const healthCheck = async (
  options?: RequestInit,
): Promise<HealthStatus> => {
  return customFetch<HealthStatus>(getHealthCheckUrl(), {
    ...options,
    method: "GET",
  });
};

export const getHealthCheckQueryKey = () => {
  return [`/api/healthz`] as const;
};

export const getHealthCheckQueryOptions = <
  TData = Awaited<ReturnType<typeof healthCheck>>,
  TError = ErrorType<unknown>,
>(options?: {
  query?: UseQueryOptions<
    Awaited<ReturnType<typeof healthCheck>>,
    TError,
    TData
  >;
  request?: SecondParameter<typeof customFetch>;
}) => {
  const { query: queryOptions, request: requestOptions } = options ?? {};

  const queryKey = queryOptions?.queryKey ?? getHealthCheckQueryKey();

  const queryFn: QueryFunction<Awaited<ReturnType<typeof healthCheck>>> = ({
    signal,
  }) => healthCheck({ signal, ...requestOptions });

  return { queryKey, queryFn, ...queryOptions } as UseQueryOptions<
    Awaited<ReturnType<typeof healthCheck>>,
    TError,
    TData
  > & { queryKey: QueryKey };
};

export type HealthCheckQueryResult = NonNullable<
  Awaited<ReturnType<typeof healthCheck>>
>;
export type HealthCheckQueryError = ErrorType<unknown>;

/**
 * @summary Health check
 */

export function useHealthCheck<
  TData = Awaited<ReturnType<typeof healthCheck>>,
  TError = ErrorType<unknown>,
>(options?: {
  query?: UseQueryOptions<
    Awaited<ReturnType<typeof healthCheck>>,
    TError,
    TData
  >;
  request?: SecondParameter<typeof customFetch>;
}): UseQueryResult<TData, TError> & { queryKey: QueryKey } {
  const queryOptions = getHealthCheckQueryOptions(options);

  const query = useQuery(queryOptions) as UseQueryResult<TData, TError> & {
    queryKey: QueryKey;
  };

  return { ...query, queryKey: queryOptions.queryKey };
}

/**
 * Accepts audio file, transcribes it, and performs cognitive analysis
 * @summary Transcribe and analyze audio
 */
export const getTranscribeAudioUrl = () => {
  return `/api/analysis/transcribe`;
};

export const transcribeAudio = async (
  transcribeAudioBody: TranscribeAudioBody,
  options?: RequestInit,
): Promise<AnalysisResult> => {
  const formData = new FormData();
  formData.append(`audio`, transcribeAudioBody.audio);
  if (transcribeAudioBody.language !== undefined) {
    formData.append(`language`, transcribeAudioBody.language);
  }
  if (
    transcribeAudioBody.userId !== undefined &&
    transcribeAudioBody.userId !== null
  ) {
    formData.append(`userId`, transcribeAudioBody.userId);
  }

  return customFetch<AnalysisResult>(getTranscribeAudioUrl(), {
    ...options,
    method: "POST",
    body: formData,
  });
};

export const getTranscribeAudioMutationOptions = <
  TError = ErrorType<void>,
  TContext = unknown,
>(options?: {
  mutation?: UseMutationOptions<
    Awaited<ReturnType<typeof transcribeAudio>>,
    TError,
    { data: BodyType<TranscribeAudioBody> },
    TContext
  >;
  request?: SecondParameter<typeof customFetch>;
}): UseMutationOptions<
  Awaited<ReturnType<typeof transcribeAudio>>,
  TError,
  { data: BodyType<TranscribeAudioBody> },
  TContext
> => {
  const mutationKey = ["transcribeAudio"];
  const { mutation: mutationOptions, request: requestOptions } = options
    ? options.mutation &&
      "mutationKey" in options.mutation &&
      options.mutation.mutationKey
      ? options
      : { ...options, mutation: { ...options.mutation, mutationKey } }
    : { mutation: { mutationKey }, request: undefined };

  const mutationFn: MutationFunction<
    Awaited<ReturnType<typeof transcribeAudio>>,
    { data: BodyType<TranscribeAudioBody> }
  > = (props) => {
    const { data } = props ?? {};

    return transcribeAudio(data, requestOptions);
  };

  return { mutationFn, ...mutationOptions };
};

export type TranscribeAudioMutationResult = NonNullable<
  Awaited<ReturnType<typeof transcribeAudio>>
>;
export type TranscribeAudioMutationBody = BodyType<TranscribeAudioBody>;
export type TranscribeAudioMutationError = ErrorType<void>;

/**
 * @summary Transcribe and analyze audio
 */
export const useTranscribeAudio = <
  TError = ErrorType<void>,
  TContext = unknown,
>(options?: {
  mutation?: UseMutationOptions<
    Awaited<ReturnType<typeof transcribeAudio>>,
    TError,
    { data: BodyType<TranscribeAudioBody> },
    TContext
  >;
  request?: SecondParameter<typeof customFetch>;
}): UseMutationResult<
  Awaited<ReturnType<typeof transcribeAudio>>,
  TError,
  { data: BodyType<TranscribeAudioBody> },
  TContext
> => {
  return useMutation(getTranscribeAudioMutationOptions(options));
};

/**
 * Analyze transcribed text for dementia risk markers
 * @summary Analyze text for cognitive markers
 */
export const getAnalyzeTextUrl = () => {
  return `/api/analysis/text`;
};

export const analyzeText = async (
  analyzeTextBody: AnalyzeTextBody,
  options?: RequestInit,
): Promise<AnalysisResult> => {
  return customFetch<AnalysisResult>(getAnalyzeTextUrl(), {
    ...options,
    method: "POST",
    headers: { "Content-Type": "application/json", ...options?.headers },
    body: JSON.stringify(analyzeTextBody),
  });
};

export const getAnalyzeTextMutationOptions = <
  TError = ErrorType<void>,
  TContext = unknown,
>(options?: {
  mutation?: UseMutationOptions<
    Awaited<ReturnType<typeof analyzeText>>,
    TError,
    { data: BodyType<AnalyzeTextBody> },
    TContext
  >;
  request?: SecondParameter<typeof customFetch>;
}): UseMutationOptions<
  Awaited<ReturnType<typeof analyzeText>>,
  TError,
  { data: BodyType<AnalyzeTextBody> },
  TContext
> => {
  const mutationKey = ["analyzeText"];
  const { mutation: mutationOptions, request: requestOptions } = options
    ? options.mutation &&
      "mutationKey" in options.mutation &&
      options.mutation.mutationKey
      ? options
      : { ...options, mutation: { ...options.mutation, mutationKey } }
    : { mutation: { mutationKey }, request: undefined };

  const mutationFn: MutationFunction<
    Awaited<ReturnType<typeof analyzeText>>,
    { data: BodyType<AnalyzeTextBody> }
  > = (props) => {
    const { data } = props ?? {};

    return analyzeText(data, requestOptions);
  };

  return { mutationFn, ...mutationOptions };
};

export type AnalyzeTextMutationResult = NonNullable<
  Awaited<ReturnType<typeof analyzeText>>
>;
export type AnalyzeTextMutationBody = BodyType<AnalyzeTextBody>;
export type AnalyzeTextMutationError = ErrorType<void>;

/**
 * @summary Analyze text for cognitive markers
 */
export const useAnalyzeText = <
  TError = ErrorType<void>,
  TContext = unknown,
>(options?: {
  mutation?: UseMutationOptions<
    Awaited<ReturnType<typeof analyzeText>>,
    TError,
    { data: BodyType<AnalyzeTextBody> },
    TContext
  >;
  request?: SecondParameter<typeof customFetch>;
}): UseMutationResult<
  Awaited<ReturnType<typeof analyzeText>>,
  TError,
  { data: BodyType<AnalyzeTextBody> },
  TContext
> => {
  return useMutation(getAnalyzeTextMutationOptions(options));
};

/**
 * @summary Get all reports for a user
 */
export const getGetReportsUrl = (params?: GetReportsParams) => {
  const normalizedParams = new URLSearchParams();

  Object.entries(params || {}).forEach(([key, value]) => {
    if (value !== undefined) {
      normalizedParams.append(key, value === null ? "null" : value.toString());
    }
  });

  const stringifiedParams = normalizedParams.toString();

  return stringifiedParams.length > 0
    ? `/api/reports?${stringifiedParams}`
    : `/api/reports`;
};

export const getReports = async (
  params?: GetReportsParams,
  options?: RequestInit,
): Promise<Report[]> => {
  return customFetch<Report[]>(getGetReportsUrl(params), {
    ...options,
    method: "GET",
  });
};

export const getGetReportsQueryKey = (params?: GetReportsParams) => {
  return [`/api/reports`, ...(params ? [params] : [])] as const;
};

export const getGetReportsQueryOptions = <
  TData = Awaited<ReturnType<typeof getReports>>,
  TError = ErrorType<unknown>,
>(
  params?: GetReportsParams,
  options?: {
    query?: UseQueryOptions<
      Awaited<ReturnType<typeof getReports>>,
      TError,
      TData
    >;
    request?: SecondParameter<typeof customFetch>;
  },
) => {
  const { query: queryOptions, request: requestOptions } = options ?? {};

  const queryKey = queryOptions?.queryKey ?? getGetReportsQueryKey(params);

  const queryFn: QueryFunction<Awaited<ReturnType<typeof getReports>>> = ({
    signal,
  }) => getReports(params, { signal, ...requestOptions });

  return { queryKey, queryFn, ...queryOptions } as UseQueryOptions<
    Awaited<ReturnType<typeof getReports>>,
    TError,
    TData
  > & { queryKey: QueryKey };
};

export type GetReportsQueryResult = NonNullable<
  Awaited<ReturnType<typeof getReports>>
>;
export type GetReportsQueryError = ErrorType<unknown>;

/**
 * @summary Get all reports for a user
 */

export function useGetReports<
  TData = Awaited<ReturnType<typeof getReports>>,
  TError = ErrorType<unknown>,
>(
  params?: GetReportsParams,
  options?: {
    query?: UseQueryOptions<
      Awaited<ReturnType<typeof getReports>>,
      TError,
      TData
    >;
    request?: SecondParameter<typeof customFetch>;
  },
): UseQueryResult<TData, TError> & { queryKey: QueryKey } {
  const queryOptions = getGetReportsQueryOptions(params, options);

  const query = useQuery(queryOptions) as UseQueryResult<TData, TError> & {
    queryKey: QueryKey;
  };

  return { ...query, queryKey: queryOptions.queryKey };
}

/**
 * @summary Save a report
 */
export const getCreateReportUrl = () => {
  return `/api/reports`;
};

export const createReport = async (
  createReportBody: CreateReportBody,
  options?: RequestInit,
): Promise<Report> => {
  return customFetch<Report>(getCreateReportUrl(), {
    ...options,
    method: "POST",
    headers: { "Content-Type": "application/json", ...options?.headers },
    body: JSON.stringify(createReportBody),
  });
};

export const getCreateReportMutationOptions = <
  TError = ErrorType<void>,
  TContext = unknown,
>(options?: {
  mutation?: UseMutationOptions<
    Awaited<ReturnType<typeof createReport>>,
    TError,
    { data: BodyType<CreateReportBody> },
    TContext
  >;
  request?: SecondParameter<typeof customFetch>;
}): UseMutationOptions<
  Awaited<ReturnType<typeof createReport>>,
  TError,
  { data: BodyType<CreateReportBody> },
  TContext
> => {
  const mutationKey = ["createReport"];
  const { mutation: mutationOptions, request: requestOptions } = options
    ? options.mutation &&
      "mutationKey" in options.mutation &&
      options.mutation.mutationKey
      ? options
      : { ...options, mutation: { ...options.mutation, mutationKey } }
    : { mutation: { mutationKey }, request: undefined };

  const mutationFn: MutationFunction<
    Awaited<ReturnType<typeof createReport>>,
    { data: BodyType<CreateReportBody> }
  > = (props) => {
    const { data } = props ?? {};

    return createReport(data, requestOptions);
  };

  return { mutationFn, ...mutationOptions };
};

export type CreateReportMutationResult = NonNullable<
  Awaited<ReturnType<typeof createReport>>
>;
export type CreateReportMutationBody = BodyType<CreateReportBody>;
export type CreateReportMutationError = ErrorType<void>;

/**
 * @summary Save a report
 */
export const useCreateReport = <
  TError = ErrorType<void>,
  TContext = unknown,
>(options?: {
  mutation?: UseMutationOptions<
    Awaited<ReturnType<typeof createReport>>,
    TError,
    { data: BodyType<CreateReportBody> },
    TContext
  >;
  request?: SecondParameter<typeof customFetch>;
}): UseMutationResult<
  Awaited<ReturnType<typeof createReport>>,
  TError,
  { data: BodyType<CreateReportBody> },
  TContext
> => {
  return useMutation(getCreateReportMutationOptions(options));
};

/**
 * @summary Get a specific report
 */
export const getGetReportUrl = (id: number) => {
  return `/api/reports/${id}`;
};

export const getReport = async (
  id: number,
  options?: RequestInit,
): Promise<Report> => {
  return customFetch<Report>(getGetReportUrl(id), {
    ...options,
    method: "GET",
  });
};

export const getGetReportQueryKey = (id: number) => {
  return [`/api/reports/${id}`] as const;
};

export const getGetReportQueryOptions = <
  TData = Awaited<ReturnType<typeof getReport>>,
  TError = ErrorType<void>,
>(
  id: number,
  options?: {
    query?: UseQueryOptions<
      Awaited<ReturnType<typeof getReport>>,
      TError,
      TData
    >;
    request?: SecondParameter<typeof customFetch>;
  },
) => {
  const { query: queryOptions, request: requestOptions } = options ?? {};

  const queryKey = queryOptions?.queryKey ?? getGetReportQueryKey(id);

  const queryFn: QueryFunction<Awaited<ReturnType<typeof getReport>>> = ({
    signal,
  }) => getReport(id, { signal, ...requestOptions });

  return {
    queryKey,
    queryFn,
    enabled: !!id,
    ...queryOptions,
  } as UseQueryOptions<Awaited<ReturnType<typeof getReport>>, TError, TData> & {
    queryKey: QueryKey;
  };
};

export type GetReportQueryResult = NonNullable<
  Awaited<ReturnType<typeof getReport>>
>;
export type GetReportQueryError = ErrorType<void>;

/**
 * @summary Get a specific report
 */

export function useGetReport<
  TData = Awaited<ReturnType<typeof getReport>>,
  TError = ErrorType<void>,
>(
  id: number,
  options?: {
    query?: UseQueryOptions<
      Awaited<ReturnType<typeof getReport>>,
      TError,
      TData
    >;
    request?: SecondParameter<typeof customFetch>;
  },
): UseQueryResult<TData, TError> & { queryKey: QueryKey } {
  const queryOptions = getGetReportQueryOptions(id, options);

  const query = useQuery(queryOptions) as UseQueryResult<TData, TError> & {
    queryKey: QueryKey;
  };

  return { ...query, queryKey: queryOptions.queryKey };
}

/**
 * @summary Delete a report
 */
export const getDeleteReportUrl = (id: number) => {
  return `/api/reports/${id}`;
};

export const deleteReport = async (
  id: number,
  options?: RequestInit,
): Promise<void> => {
  return customFetch<void>(getDeleteReportUrl(id), {
    ...options,
    method: "DELETE",
  });
};

export const getDeleteReportMutationOptions = <
  TError = ErrorType<void>,
  TContext = unknown,
>(options?: {
  mutation?: UseMutationOptions<
    Awaited<ReturnType<typeof deleteReport>>,
    TError,
    { id: number },
    TContext
  >;
  request?: SecondParameter<typeof customFetch>;
}): UseMutationOptions<
  Awaited<ReturnType<typeof deleteReport>>,
  TError,
  { id: number },
  TContext
> => {
  const mutationKey = ["deleteReport"];
  const { mutation: mutationOptions, request: requestOptions } = options
    ? options.mutation &&
      "mutationKey" in options.mutation &&
      options.mutation.mutationKey
      ? options
      : { ...options, mutation: { ...options.mutation, mutationKey } }
    : { mutation: { mutationKey }, request: undefined };

  const mutationFn: MutationFunction<
    Awaited<ReturnType<typeof deleteReport>>,
    { id: number }
  > = (props) => {
    const { id } = props ?? {};

    return deleteReport(id, requestOptions);
  };

  return { mutationFn, ...mutationOptions };
};

export type DeleteReportMutationResult = NonNullable<
  Awaited<ReturnType<typeof deleteReport>>
>;

export type DeleteReportMutationError = ErrorType<void>;

/**
 * @summary Delete a report
 */
export const useDeleteReport = <
  TError = ErrorType<void>,
  TContext = unknown,
>(options?: {
  mutation?: UseMutationOptions<
    Awaited<ReturnType<typeof deleteReport>>,
    TError,
    { id: number },
    TContext
  >;
  request?: SecondParameter<typeof customFetch>;
}): UseMutationResult<
  Awaited<ReturnType<typeof deleteReport>>,
  TError,
  { id: number },
  TContext
> => {
  return useMutation(getDeleteReportMutationOptions(options));
};

/**
 * @summary Send message to cognitive health chatbot
 */
export const getSendChatbotMessageUrl = () => {
  return `/api/chatbot/message`;
};

export const sendChatbotMessage = async (
  chatbotMessageBody: ChatbotMessageBody,
  options?: RequestInit,
): Promise<ChatbotResponse> => {
  return customFetch<ChatbotResponse>(getSendChatbotMessageUrl(), {
    ...options,
    method: "POST",
    headers: { "Content-Type": "application/json", ...options?.headers },
    body: JSON.stringify(chatbotMessageBody),
  });
};

export const getSendChatbotMessageMutationOptions = <
  TError = ErrorType<void>,
  TContext = unknown,
>(options?: {
  mutation?: UseMutationOptions<
    Awaited<ReturnType<typeof sendChatbotMessage>>,
    TError,
    { data: BodyType<ChatbotMessageBody> },
    TContext
  >;
  request?: SecondParameter<typeof customFetch>;
}): UseMutationOptions<
  Awaited<ReturnType<typeof sendChatbotMessage>>,
  TError,
  { data: BodyType<ChatbotMessageBody> },
  TContext
> => {
  const mutationKey = ["sendChatbotMessage"];
  const { mutation: mutationOptions, request: requestOptions } = options
    ? options.mutation &&
      "mutationKey" in options.mutation &&
      options.mutation.mutationKey
      ? options
      : { ...options, mutation: { ...options.mutation, mutationKey } }
    : { mutation: { mutationKey }, request: undefined };

  const mutationFn: MutationFunction<
    Awaited<ReturnType<typeof sendChatbotMessage>>,
    { data: BodyType<ChatbotMessageBody> }
  > = (props) => {
    const { data } = props ?? {};

    return sendChatbotMessage(data, requestOptions);
  };

  return { mutationFn, ...mutationOptions };
};

export type SendChatbotMessageMutationResult = NonNullable<
  Awaited<ReturnType<typeof sendChatbotMessage>>
>;
export type SendChatbotMessageMutationBody = BodyType<ChatbotMessageBody>;
export type SendChatbotMessageMutationError = ErrorType<void>;

/**
 * @summary Send message to cognitive health chatbot
 */
export const useSendChatbotMessage = <
  TError = ErrorType<void>,
  TContext = unknown,
>(options?: {
  mutation?: UseMutationOptions<
    Awaited<ReturnType<typeof sendChatbotMessage>>,
    TError,
    { data: BodyType<ChatbotMessageBody> },
    TContext
  >;
  request?: SecondParameter<typeof customFetch>;
}): UseMutationResult<
  Awaited<ReturnType<typeof sendChatbotMessage>>,
  TError,
  { data: BodyType<ChatbotMessageBody> },
  TContext
> => {
  return useMutation(getSendChatbotMessageMutationOptions(options));
};
