// @workspace/.ignored_api-client-react/src/generated/api.schemas.ts
/**
 * Generated by orval v8.5.3 🍺
 * Do not edit manually.
 * Api
 * Early Dementia Detection via Speech Analysis API
 * OpenAPI spec version: 0.1.0
 */
export interface HealthStatus {
  status: string;
}

export interface SpeechFeatures {
  wordCount: number;
  uniqueWordCount: number;
  vocabularyDiversity: number;
  avgSentenceLength: number;
  pauseCount: number;
  repetitionCount: number;
  repetitionRate: number;
  fillerWordCount: number;
  coherenceScore: number;
}

export type AnalysisResultRiskLevel =
  (typeof AnalysisResultRiskLevel)[keyof typeof AnalysisResultRiskLevel];

export const AnalysisResultRiskLevel = {
  Normal: "Normal",
  Mild_Cognitive_Impairment: "Mild Cognitive Impairment",
  High_Risk: "High Risk",
} as const;

export interface AnalysisResult {
  transcript: string;
  riskLevel: AnalysisResultRiskLevel;
  riskScore: number;
  confidence: number;
  features: SpeechFeatures;
  problematicPatterns: string[];
  recommendations: string[];
  language: string;
}

export type AnalyzeTextBodyLanguage =
  (typeof AnalyzeTextBodyLanguage)[keyof typeof AnalyzeTextBodyLanguage];

export const AnalyzeTextBodyLanguage = {
  en: "en",
  te: "te",
} as const;

export interface AnalyzeTextBody {
  text: string;
  language?: AnalyzeTextBodyLanguage;
  /** @nullable */
  userId?: string | null;
}

export interface Report {
  id: number;
  userId: string;
  transcript: string;
  riskLevel: string;
  riskScore: number;
  confidence: number;
  features: SpeechFeatures;
  problematicPatterns: string[];
  recommendations: string[];
  language: string;
  createdAt: string;
}

export interface CreateReportBody {
  userId: string;
  transcript: string;
  riskLevel: string;
  riskScore: number;
  confidence: number;
  features: SpeechFeatures;
  problematicPatterns: string[];
  recommendations: string[];
  language: string;
}

export type ChatMessageRole =
  (typeof ChatMessageRole)[keyof typeof ChatMessageRole];

export const ChatMessageRole = {
  user: "user",
  assistant: "assistant",
} as const;

export interface ChatMessage {
  role: ChatMessageRole;
  content: string;
}

export type ChatbotMessageBodyLanguage =
  (typeof ChatbotMessageBodyLanguage)[keyof typeof ChatbotMessageBodyLanguage];

export const ChatbotMessageBodyLanguage = {
  en: "en",
  te: "te",
} as const;

export interface ChatbotMessageBody {
  message: string;
  history?: ChatMessage[];
  language?: ChatbotMessageBodyLanguage;
}

export interface ChatbotResponse {
  reply: string;
  tips: string[];
}

export type TranscribeAudioBodyLanguage =
  (typeof TranscribeAudioBodyLanguage)[keyof typeof TranscribeAudioBodyLanguage];

export const TranscribeAudioBodyLanguage = {
  en: "en",
  te: "te",
} as const;

export type TranscribeAudioBody = {
  audio: Blob;
  language?: TranscribeAudioBodyLanguage;
  /** @nullable */
  userId?: string | null;
};

export type GetReportsParams = {
  userId?: string;
};
